relevanceai/operations_new/cluster/base.py

```diff
@@ -32,22 +84,38 @@ def run(self, documents: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
 
         """
 
-        updated_documents = documents
+        updated_documents = deepcopy(documents)
```

Comment:
The performance of this code needs to be looked at.

A copy is not necessary here.

Reply:
@JackyKoh 

A copy is necessary because if we cache the documents once they are retrieved, we don't want to edit those cached documents and re-upload them separately.

Reply:
This is how you do it:
- self.model.fit_predict_documents doesn't update the original documents

step 1: after fit_predict_documents, take documents only for their ids
`documents_to_upsert = [{"_id": d["_id"]} for d in documents]`

step 2: `self.set_field_across_documents(..., documents_to_upsert )`